src/Features/Core/Portable/CodeFixesAndRefactorings/CodeActionRequestPriorityProvider.cs

```diff
@@ -0,0 +1,107 @@
+        /// Default provider with <see cref="CodeActionRequestPriority.None"/> <see cref="Priority"/>.
+        /// </summary>
+        public static readonly CodeActionRequestPriorityProvider Default = DefaultProvider.Instance;
+
+        protected CodeActionRequestPriorityProvider(CodeActionRequestPriority priority)
+        {
+            Priority = priority;
+        }
+
+        public CodeActionRequestPriority Priority { get; }
+
+        protected abstract bool IsDeprioritizedAnalyzerWithLowPriority(DiagnosticAnalyzer analyzer);
+
+        /// <summary>
+        /// Tracks the given <paramref name="analyzer"/> as a de-prioritized analyzer that should be moved to
+        /// <see cref="CodeActionRequestPriority.Low"/> bucket.
+        /// </summary>
+        public abstract void AddDeprioritizedAnalyzerWithLowPriority(DiagnosticAnalyzer analyzer);
+
+        public abstract CodeActionRequestPriorityProvider With(CodeActionRequestPriority priority);
+
+        public static CodeActionRequestPriorityProvider Create(CodeActionRequestPriority priority)
+        {
+            if (priority == CodeActionRequestPriority.None)
+                return Default;
+
+            return MutableProvider.Create(priority);
+        }
+
+        /// <summary>
+        /// Returns true if the given <paramref name="analyzer"/> can report diagnostics that can have
+        /// fixes from a code fix provider with <see cref="CodeFixProvider.RequestPriority"/>
+        /// matching <see cref="Priority"/>.
+        /// This method is useful for performing a performance optimization for lightbulb diagnostic computation,
+        /// wherein we can reduce the set of analyzers to be executed when computing fixes for a specific
+        /// <see cref="Priority"/>.
+        /// </summary>
+        public bool MatchesPriority(DiagnosticAnalyzer analyzer)
+        {
+            // If caller isn't asking for prioritized result, then run all analyzers.
+            if (Priority == CodeActionRequestPriority.None)
+                return true;
+
+            // 'CodeActionRequestPriority.Lowest' is used for suppression/configuration fixes,
+            // which requires all analyzer diagnostics.
+            if (Priority == CodeActionRequestPriority.Lowest)
+                return true;
+
+            // The compiler analyzer always counts for any priority.  It's diagnostics may be fixed
+            // by high pri or normal pri fixers.
+            if (analyzer.IsCompilerAnalyzer())
+                return true;
+
+            // Check if we are computing diagnostics for 'CodeActionRequestPriority.Low' and
+            // this analyzer was de-prioritized to low priority bucket.
+            if (Priority == CodeActionRequestPriority.Low &&
+                IsDeprioritizedAnalyzerWithLowPriority(analyzer))
+            {
+                return true;
+            }
+
+            var analyzerPriority = analyzer is IBuiltInAnalyzer { RequestPriority: var requestPriority }
```

Comment:
i would still doc this branch.  but thanks for all the docs above.  it was very helpful.